src/osd/scheduler/mClockScheduler.cc

```diff
@@ -30,13 +30,98 @@ using namespace std::placeholders;
 
 namespace ceph::osd::scheduler {
 
+void mClockScheduler::_get_mclock_counter(scheduler_id_t id)
+{
+  if (!logger) {
+    return;
+  }
+
+  /* op enter mclock queue will +1 */
+  logger->inc(l_mclock_all_type_queue_len);
+
+  switch (id.class_id) {
+  case op_scheduler_class::immediate:
+    logger->inc(l_mclock_immediate_queue_len);
+    break;
+  case op_scheduler_class::client:
+    logger->inc(l_mclock_client_queue_len);
+    break;
+  case op_scheduler_class::background_recovery:
+    logger->inc(l_mclock_recovery_queue_len);
+    break;
+  case op_scheduler_class::background_best_effort:
+    logger->inc(l_mclock_best_effort_queue_len);
+    break;
+   default:
+    derr << __func__ << " unknown class_id=" << id.class_id
+         << " unknown id=" << id << dendl;
+    break;
+  }
+}
+
+void mClockScheduler::_put_mclock_counter(scheduler_id_t id)
+{
+  if (!logger) {
+    return;
+  }
+
+  /* op leave mclock queue will -1 */
+  logger->dec(l_mclock_all_type_queue_len);
+
+  switch (id.class_id) {
+  case op_scheduler_class::immediate:
+    logger->dec(l_mclock_immediate_queue_len);
+    break;
+  case op_scheduler_class::client:
+    logger->dec(l_mclock_client_queue_len);
+    break;
+  case op_scheduler_class::background_recovery:
+    logger->dec(l_mclock_recovery_queue_len);
+    break;
+  case op_scheduler_class::background_best_effort:
+    logger->dec(l_mclock_best_effort_queue_len);
+    break;
+   default:
+    derr << __func__ << " unknown class_id=" << id.class_id
+         << " unknown id=" << id << dendl;
+    break;
+  }
+}
+
+void mClockScheduler::_init_logger()
+{
+  PerfCountersBuilder m(cct, "mclock-shard-queue-" + std::to_string(shard_id),
+                        l_mclock_first, l_mclock_last);
+
+  m.add_u64_counter(l_mclock_immediate_queue_len, "mclock_immediate_queue_len",
+                    "high_priority op count in mclock queue");
+  m.add_u64_counter(l_mclock_client_queue_len, "mclock_client_queue_len",
+                    "client type op count in mclock queue");
+  m.add_u64_counter(l_mclock_recovery_queue_len, "mclock_recovery_queue_len",
+                    "background_recovery type op count in mclock queue");
+  m.add_u64_counter(l_mclock_best_effort_queue_len, "mclock_best_effort_queue_len",
+                    "background_best_effort type op count in mclock queue");
+  m.add_u64_counter(l_mclock_all_type_queue_len, "mclock_all_type_queue_len",
+                    "all type op count in mclock queue");
+
+  logger = m.create_perf_counters();
+  cct->get_perfcounters_collection()->add(logger);
```

Comment:
**unittest_mclock_scheduler**  failed at here,  while 1 loop:

In unittest_mclock_scheduler, 
- multiple test cases use the same logger ADDR 
- and add it to the logger collection of the same g_ceph_context cct
- Caused a dead loop

Solution:
- I don't think it's mclock's fault.
- I think it's a problem with perfcounter itself (add/remove...)
- This PR does not introduce this issue, It should be introduced by e9ecd1b38447  @sseshasa 
- So I added ignore_perfcounter=true parameter to TEST

https://github.com/ceph/ceph/blob/main/src/common/perf_counters.cc#L41
```
void PerfCountersCollectionImpl::add(PerfCounters *l)
{
  // make sure the name is unique
  perf_counters_set_t::iterator i;
  i = m_loggers.find(l);
  while (i != m_loggers.end()) {
    ostringstream ss;
    ss << l->get_name() << "-" << (void*)l;
    l->set_name(ss.str());
    i = m_loggers.find(l);
  }

  m_loggers.insert(l);
  ... ...
}
```

https://github.com/ceph/ceph/blob/main/src/common/perf_counters.h#L359
https://github.com/ceph/ceph/blob/main/src/common/perf_counters.h#L312
```
class SortPerfCountersByName {
public:
  bool operator()(const PerfCounters* lhs, const PerfCounters* rhs) const {
    return (lhs->get_name() < rhs->get_name());
  }
};
 
typedef std::set <PerfCounters*, SortPerfCountersByName> perf_counters_set_t;
 
//PerfCountersCollectionImpl
  perf_counters_set_t m_loggers;
```

```
[----------] 5 tests from mClockSchedulerTest
[ RUN      ] mClockSchedulerTest.TestEmpty
_init_logger zhangjianwei2 mclock-shard-queue-0
_init_logger zhangjianwei2 logger=0x5610ac84e0a0  //same addr
add zhangjianwei2 111111
add zhangjianwei2 121111 l=0x5610ac84e0a0          //first time add to logger collection
add zhangjianwei2 131111
add zhangjianwei2 141111
add zhangjianwei2 151111
_init_logger zhangjianwei2 add logger=0x5610ac84e0a0  
[       OK ] mClockSchedulerTest.TestEmpty (0 ms)
[ RUN      ] mClockSchedulerTest.TestSingleClientOrderedEnqueueDequeue
_init_logger zhangjianwei2 mclock-shard-queue-0
_init_logger zhangjianwei2 logger=0x5610ac84e0a0   //same addr
add zhangjianwei2 111111                                            //PerfCountersCollectionImpl::add
add zhangjianwei2 121111 l=0x5610ac84e0a0           //second time add to logger collection, dead loop here 
add zhangjianwei2 120111 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0
add zhangjianwei2 120211 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0
add zhangjianwei2 120111 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0-0x5610ac84e0a0
add zhangjianwei2 120211 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0-0x5610ac84e0a0
add zhangjianwei2 120111 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0-0x5610ac84e0a0-0x5610ac84e0a0
add zhangjianwei2 120211 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0-0x5610ac84e0a0-0x5610ac84e0a0
add zhangjianwei2 120111 l=0x5610ac84e0a0 mclock-shard-queue-0-0x5610ac84e0a0-0x5610ac84e0a0-0x5610ac84e0a0-0x5610ac84e0a0
```

Reply:
create issue: src/common: dead loop in PerfCountersCollectionImpl::add() when adding the same logger_ptr to logger_collection https://tracker.ceph.com/issues/66758